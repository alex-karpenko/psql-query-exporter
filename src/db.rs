use serde::Deserialize;
use std::{fmt::Display, time::Duration};
use tracing::{debug, error};

use openssl::ssl::{SslConnector, SslMethod, SslVerifyMode};
use postgres_openssl::MakeTlsConnector;
use tokio::task::JoinHandle;
use tokio_postgres::{Client, Error, Row};

#[derive(Debug)]
pub struct PostgresConnection {
    db_connection_string: String,
    client: Client,
    connection_handler: JoinHandle<()>,
    sslmode: PostgresSslMode,
    ssl_rootcert: Option<String>,
    default_backoff_interval: Duration,
    max_backoff_interval: Duration,
}

#[derive(Deserialize, Debug, Clone, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum PostgresSslMode {
    Disable,
    Prefer,
    Require,
    #[serde(rename = "verify-ca")]
    VerifyCa,
    #[serde(rename = "verify-full")]
    VerifyFull,
}

impl Default for PostgresSslMode {
    fn default() -> Self {
        Self::Prefer
    }
}

impl Display for PostgresSslMode {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let s = match self {
            Self::Disable => "disable",
            Self::Prefer => "prefer",
            Self::Require => "require",
            Self::VerifyCa => "require",
            Self::VerifyFull => "require",
        };
        write!(f, "{s}")
    }
}

impl PostgresSslMode {
    fn get_mode_str(&self) -> &str {
        match self {
            PostgresSslMode::Disable => "Disabled",
            PostgresSslMode::Prefer => "Prefer",
            PostgresSslMode::Require => "Require",
            PostgresSslMode::VerifyCa => "Verify-CA",
            PostgresSslMode::VerifyFull => "Verify-Full",
        }
    }
}

impl PostgresConnection {
    pub async fn new(
        db_connection_string: String,
        sslmode: PostgresSslMode,
        ssl_rootcert: Option<String>,
        default_backoff_interval: Duration,
        max_backoff_interval: Duration,
    ) -> Result<Self, Error> {
        debug!("PostgresConnection::new: construct new postgres connection");

        let mut backoff_interval = default_backoff_interval;

        loop {
            debug!("sslmode={}", sslmode.get_mode_str());
            let mut connector = SslConnector::builder(SslMethod::tls())
                .unwrap_or_else(|e| panic!("unable to create SSL connector: {}", e));

            match sslmode {
                PostgresSslMode::Disable => connector.set_verify(SslVerifyMode::NONE),
                PostgresSslMode::Prefer => connector.set_verify(SslVerifyMode::NONE),
                PostgresSslMode::Require => connector.set_verify(SslVerifyMode::NONE),
                PostgresSslMode::VerifyCa => {
                    connector.set_verify_callback(
                        SslVerifyMode::PEER,
                        |verify_indicator, x509_result| {
                            let allowed_errors: Vec<i32> = vec![
                                openssl_sys::X509_V_ERR_IP_ADDRESS_MISMATCH,
                                openssl_sys::X509_V_ERR_HOSTNAME_MISMATCH,
                                openssl_sys::X509_V_ERR_EMAIL_MISMATCH,
                            ];
                            debug!(
                                "verify_callback, indicator={}, x509_result={}",
                                verify_indicator,
                                x509_result.error()
                            );

                            if !verify_indicator
                                && allowed_errors.contains(&x509_result.error().as_raw())
                            {
                                true
                            } else {
                                verify_indicator
                            }
                        },
                    );
                }
                PostgresSslMode::VerifyFull => connector.set_verify(SslVerifyMode::PEER),
            };

            if let Some(rootcert) = ssl_rootcert.as_ref() {
                debug!("loading CA bundle from {}", rootcert);
                connector
                    .set_ca_file(rootcert)
                    .unwrap_or_else(|e| panic!("unable to load PEM CA file {}: {}", rootcert, e));
            }

            let connector = MakeTlsConnector::new(connector.build());
            let connection = tokio_postgres::connect(&db_connection_string, connector).await;

            match connection {
                Ok((client, connection)) => {
                    let connection_handler = tokio::spawn(async move {
                        debug!("PostgresConnection::new: spawn new connection task");
                        if let Err(e) = connection.await {
                            error!("PostgresConnection: connection closed with error: {}", e);
                        }
                    });

                    return Ok(PostgresConnection {
                        client,
                        db_connection_string,
                        connection_handler,
                        sslmode,
                        ssl_rootcert,
                        default_backoff_interval,
                        max_backoff_interval,
                    });
                }
                Err(e) => {
                    error!("PostgresConnection::new: client error: {e}");
                }
            };

            tokio::time::sleep(backoff_interval).await;
            backoff_interval += default_backoff_interval;
            if backoff_interval > max_backoff_interval {
                backoff_interval = max_backoff_interval;
            }
        }
    }

    async fn reconnect(&mut self) -> Result<&Self, Error> {
        debug!("PostgresConnection::reconnect: try to reconnect");
        let new_connection = PostgresConnection::new(
            self.db_connection_string.clone(),
            self.sslmode.clone(),
            self.ssl_rootcert.clone(),
            self.default_backoff_interval,
            self.max_backoff_interval,
        )
        .await;

        match new_connection {
            Ok(conn) => {
                self.client = conn.client;
                self.connection_handler = conn.connection_handler;
                Ok(self)
            }
            Err(e) => {
                error!("PostgresConnection::reconnect: can't reconnect: {e}");
                Err(e)
            }
        }
    }

    pub async fn query(&mut self, query: &str, query_timeout: Duration) -> Result<Vec<Row>, Error> {
        debug!("PostgresConnection::query: {query:?}");
        let mut backoff_interval = self.default_backoff_interval;

        loop {
            // Set statement timeout
            let result = self
                .client
                .query(
                    format!("set statement_timeout={};", query_timeout.as_millis()).as_str(),
                    &[],
                )
                .await;
            if let Err(e) = result {
                error!("PostgresConnection::query: {e}");
                if e.code().is_none() {
                    debug!("PostgresConnection::query: try to reconnect after error");
                    self.reconnect().await?;
                } else {
                    return Err(e);
                }
            } else {
                // Execute actual query
                let result = self.client.query(query, &[]).await;
                if let Err(e) = result {
                    error!("PostgresConnection::query: {e}");
                    if e.code().is_none() {
                        debug!("PostgresConnection::query: try to reconnect after error");
                        self.reconnect().await?;
                    } else {
                        return Err(e);
                    }
                } else {
                    return result;
                }
            }

            tokio::time::sleep(backoff_interval).await;
            backoff_interval += self.default_backoff_interval;
            if backoff_interval > self.max_backoff_interval {
                backoff_interval = self.max_backoff_interval;
            }
        }
    }
}
