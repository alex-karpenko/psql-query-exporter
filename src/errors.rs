use std::{env, io};
use thiserror::Error;

#[derive(Error)]
pub enum PsqlExporterError {
    #[error("unable to load config file '{}': {}", .filename, .cause)]
    LoadConfigFile { filename: String, cause: io::Error },
    #[error("unable to parse config: {}", .cause.kind)]
    ParseConfigFile {
        #[from]
        cause: figment::Error,
    },
    #[error("unable to substitute environment variable '{}': {}", .variable, .cause)]
    EnvironmentVariableSubstitution {
        variable: String,
        cause: env::VarError,
    },
    #[error("query failed '{}': {}", .query, .cause)]
    PostgresQuery {
        query: String,
        cause: tokio_postgres::Error,
    },
    #[error("unable to create TLS connector: {}", .0)]
    PostgresTlsConnector(openssl::error::ErrorStack),
    #[error("unable to load CA certificate '{}': {}", .rootcert, .cause)]
    PostgresTlsRootCertificate {
        rootcert: String,
        cause: openssl::error::ErrorStack,
    },
    #[error("unable to load client certificate/key '{}': {}", .filename, .cause)]
    PostgresTlsClientCertificate {
        filename: String,
        cause: openssl::error::ErrorStack,
    },
    #[error("TLS client config error: {}", .0)]
    PostgresTlsClientConfig(String),
    #[error("shutdown signal has been received during operation")]
    ShutdownSignalReceived,
    #[error("unable to create metric '{}': {}", .metric, .cause)]
    CreateMetric {
        metric: String,
        cause: prometheus::Error,
    },
    #[error("unable to send task completion status: {}", .0)]
    MetricsBackStatusSend(#[from] tokio::sync::mpsc::error::SendError<usize>),
}

impl std::fmt::Debug for PsqlExporterError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self)
    }
}
