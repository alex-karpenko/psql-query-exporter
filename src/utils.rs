use crate::errors::PsqlExporterError;
use std::{error::Error, time::Duration};
use tokio::{
    select,
    signal::unix::{signal, Signal, SignalKind},
    sync::watch,
};
use tracing::{debug, error, info, instrument};
pub type ShutdownReceiver = watch::Receiver<bool>;
pub type ShutdownSender = watch::Sender<bool>;

#[derive(Debug)]
pub struct SignalHandler {
    terminate: Signal,
    interrupt: Signal,
    quit: Signal,
    hangup: Signal,

    shutdown_channel_tx: ShutdownSender,
    shutdown_channel_rx: ShutdownReceiver,
}

impl SignalHandler {
    pub fn new() -> Result<Self, Box<dyn Error>> {
        let (shutdown_channel_tx, shutdown_channel_rx) = watch::channel(false);
        let receiver = Self {
            terminate: signal(SignalKind::terminate())?,
            interrupt: signal(SignalKind::interrupt())?,
            quit: signal(SignalKind::quit())?,
            hangup: signal(SignalKind::hangup())?,
            shutdown_channel_tx,
            shutdown_channel_rx,
        };

        Ok(receiver)
    }

    pub fn get_rx_channel(&self) -> ShutdownReceiver {
        self.shutdown_channel_rx.clone()
    }

    #[instrument("SignalHandler", skip_all)]
    pub async fn shutdown_on_signal(&mut self) {
        let signal = self.wait_for_signal().await;

        info!(%signal,  "shutting down");
        if let Err(e) = self.shutdown_channel_tx.send(true) {
            error!(error = %e, "can't send shutdown message");
        } else {
            debug!("shutdown message has been sent, waiting until all task stopped");
            self.shutdown_channel_tx.closed().await;
            debug!("shutdown completed");
        }
    }

    async fn wait_for_signal(&mut self) -> &str {
        select! {
            _ = self.terminate.recv() => "TERM",
            _ = self.interrupt.recv() => "INT",
            _ = self.quit.recv() => "QUIT",
            _ = self.hangup.recv() => "HANGUP",
        }
    }
}

pub struct SleepHelper {
    shutdown_channel: ShutdownReceiver,
}

impl SleepHelper {
    pub fn from(shutdown_channel: ShutdownReceiver) -> Self {
        Self { shutdown_channel }
    }

    #[instrument("SleepHandler", skip_all, fields(duration=?sleep_time))]
    pub async fn sleep(&mut self, sleep_time: Duration) -> Result<(), PsqlExporterError> {
        select! {
            _ = self.shutdown_channel.changed() => {
                debug!("shutdown signal has been received");
                return Err(PsqlExporterError::ShutdownSignalReceived);
            },
            _ = tokio::time::sleep(sleep_time) => {
                debug!("timeout has been reached");
            }
        }

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_sleep_normal_timeout() {
        let (_tx, rx) = watch::channel(false);
        let mut helper = SleepHelper::from(rx);
        let sleep_duration = Duration::from_millis(100);

        select! {
            result = helper.sleep(sleep_duration) => {
                assert!(result.is_ok());
            },
            _ = tokio::time::sleep(Duration::from_millis(200)) => {
                panic!("timeout has been reached");
            }
        }
    }

    #[tokio::test]
    async fn test_sleep_shutdown_signal() {
        let (tx, rx) = watch::channel(false);
        let mut helper = SleepHelper::from(rx);
        let sleep_duration = Duration::from_secs(1);

        // Send shutdown signal
        tokio::spawn(async move {
            tokio::time::sleep(Duration::from_millis(50)).await;
            tx.send(true).unwrap();
        });

        let result = helper.sleep(sleep_duration).await;
        assert!(matches!(
            result,
            Err(PsqlExporterError::ShutdownSignalReceived)
        ));
    }
}
