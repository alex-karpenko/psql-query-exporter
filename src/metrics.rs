use crate::db::{PostgresConnection, PostgresSslCertificates};
use crate::errors::PsqlExporterError;
use crate::scrape_config::{
    FieldType, ScrapeConfig, ScrapeConfigDatabase, ScrapeConfigQuery, ScrapeConfigValues,
};
use crate::utils::{ShutdownReceiver, SleepHelper};

use prometheus::core::{AtomicF64, AtomicI64, Collector, GenericGauge, GenericGaugeVec};
use prometheus::{
    opts, Encoder, Gauge, GaugeVec, IntGauge, IntGaugeVec, Opts, Registry, TextEncoder,
};
use tokio::sync::mpsc;
use tokio_postgres::Row;

use std::convert::Infallible;
use std::time::{Duration, SystemTime};

use tracing::{debug, error, info};

#[derive(Debug)]
pub enum MetricWithType {
    SingleInt(GenericGauge<AtomicI64>),
    SingleFloat(GenericGauge<AtomicF64>),
    VectorInt(GenericGaugeVec<AtomicI64>),
    VectorFloat(GenericGaugeVec<AtomicF64>),
}

impl MetricWithType {
    fn to_collector(&self) -> Box<dyn Collector> {
        match self {
            MetricWithType::SingleInt(m) => Box::new(m.to_owned()),
            MetricWithType::SingleFloat(m) => Box::new(m.to_owned()),
            MetricWithType::VectorInt(m) => Box::new(m.to_owned()),
            MetricWithType::VectorFloat(m) => Box::new(m.to_owned()),
        }
    }
}

struct QueryMetrics {
    metrics: Vec<MetricWithType>,
    is_registered: bool,
    last_updated: SystemTime,
    next_query_time: SystemTime,
}

impl QueryMetrics {
    fn from(query_config: &ScrapeConfigQuery) -> Result<Self, PsqlExporterError> {
        let mut metrics: Vec<MetricWithType> = vec![];

        match &query_config.values {
            ScrapeConfigValues::ValueFrom(values) => {
                let mut opts = opts!(
                    query_config.metric_name.clone(),
                    query_config.description.clone().unwrap()
                );

                if let Some(const_labels) = &query_config.const_labels {
                    opts = opts.const_labels(const_labels.clone());
                }

                let new_metric =
                    Self::helper_create_metric(&query_config.var_labels, &values.field_type, opts)
                        .map_err(|e| PsqlExporterError::CreateMetric {
                            metric: query_config.metric_name.clone(),
                            cause: e,
                        })?;

                metrics.push(new_metric);
            }

            ScrapeConfigValues::ValuesWithLabels(values) => {
                for value in values {
                    let mut opts = opts!(
                        query_config.metric_name.clone(),
                        query_config.description.clone().unwrap()
                    );

                    if let Some(const_labels) = &query_config.const_labels {
                        let mut const_labels = const_labels.clone();
                        value.labels.iter().for_each(|(k, v)| {
                            const_labels.insert(k.to_string(), v.to_string());
                        });
                        opts = opts.const_labels(const_labels);
                    }
                    let new_metric = Self::helper_create_metric(
                        &query_config.var_labels,
                        &value.field_type,
                        opts,
                    )
                    .map_err(|e| PsqlExporterError::CreateMetric {
                        metric: query_config.metric_name.clone(),
                        cause: e,
                    })?;

                    metrics.push(new_metric);
                }
            }

            ScrapeConfigValues::ValuesWithSuffixes(values) => {
                for value in values {
                    let metric_name = format!("{}_{}", query_config.metric_name, value.suffix);
                    let metric_desc = format!(
                        "{}: {}",
                        query_config.description.clone().unwrap(),
                        value.suffix
                    );
                    let mut opts = opts!(metric_name, metric_desc);

                    if let Some(const_labels) = &query_config.const_labels {
                        opts = opts.const_labels(const_labels.clone());
                    }
                    let new_metric = Self::helper_create_metric(
                        &query_config.var_labels,
                        &value.field_type,
                        opts,
                    )
                    .map_err(|e| PsqlExporterError::CreateMetric {
                        metric: query_config.metric_name.clone(),
                        cause: e,
                    })?;

                    metrics.push(new_metric);
                }
            }
        };

        Ok(QueryMetrics {
            metrics,
            is_registered: false,
            last_updated: SystemTime::now() - query_config.metric_expiration_time,
            next_query_time: SystemTime::now(),
        })
    }

    fn helper_create_metric(
        var_labels: &Option<Vec<String>>,
        field_type: &FieldType,
        opts: Opts,
    ) -> Result<MetricWithType, prometheus::Error> {
        if let Some(var_labels) = var_labels {
            let new_labels: Vec<&str> = var_labels.iter().map(AsRef::as_ref).collect();
            match field_type {
                FieldType::Int => Ok(MetricWithType::VectorInt(IntGaugeVec::new(
                    opts,
                    &new_labels,
                )?)),
                FieldType::Float => Ok(MetricWithType::VectorFloat(GaugeVec::new(
                    opts,
                    &new_labels,
                )?)),
            }
        } else {
            match field_type {
                FieldType::Int => Ok(MetricWithType::SingleInt(IntGauge::with_opts(opts)?)),
                FieldType::Float => Ok(MetricWithType::SingleFloat(Gauge::with_opts(opts)?)),
            }
        }
    }

    fn register(&mut self, registry: &Registry) {
        self.last_updated = SystemTime::now();
        if !self.is_registered {
            for metric in self.metrics.iter() {
                let metric = metric.to_collector();
                registry
                    .register(metric)
                    .unwrap_or_else(|e| panic!("error while registering metric: {e}"));
            }
            self.is_registered = true;
        };
    }

    fn unregister(&mut self, registry: &Registry) {
        if self.is_registered {
            for metric in self.metrics.iter() {
                let metric = metric.to_collector();
                registry
                    .unregister(metric)
                    .unwrap_or_else(|e| panic!("error while un-registering metric: {e}"));
            }
            self.is_registered = false;
        };
    }
}

pub async fn compose_reply() -> Result<impl warp::Reply, Infallible> {
    let registry = prometheus::default_registry();
    debug!("compose_reply: preparing metrics, registry={registry:?}");

    let mut buffer = vec![];
    let encoder = TextEncoder::new();
    let metric_families = registry.gather();
    encoder
        .encode(&metric_families, &mut buffer)
        .unwrap_or_else(|e| panic!("looks like a BUG: {e}"));

    Ok(String::from_utf8(buffer).unwrap_or_else(|e| panic!("looks like a BUG: {e}")))
}

pub async fn collecting_task(
    scrape_config: ScrapeConfig,
    shutdown_channel: ShutdownReceiver,
) -> Result<(), PsqlExporterError> {
    debug!("collecting_task: config={scrape_config:?}");
    let mut handler_index: usize = 0;
    let (tx, mut rx) = mpsc::channel(scrape_config.len());
    let sources = scrape_config.sources;
    for (_, source_db_instance) in sources {
        let databases = source_db_instance.databases;
        for database in databases {
            let tx = tx.clone();
            let shut_rx = shutdown_channel.clone();
            tokio::spawn(async move {
                let handler_result = collect_one_db_instance(database, shut_rx).await;
                let send_result = tx
                    .send(handler_index)
                    .await
                    .map_err(PsqlExporterError::MetricsBackStatusSend);

                if let Err(result) = handler_result {
                    match result {
                        PsqlExporterError::ShutdownSignalReceived => {
                            debug!("collect db task #{handler_index} completed by shutdown signal");
                            Ok(())
                        }
                        _ => {
                            error!("collect db task completed unexpectedly: {result}");
                            Err(result)
                        }
                    }
                } else if let Err(result) = send_result {
                    Err(result)
                } else {
                    handler_result
                }
            });
            handler_index += 1;
        }
    }

    debug!("collecting_task: {handler_index} handlers have been started");

    while let Some(task_index) = rx.recv().await {
        debug!("collecting_task: collecting_task_handler #{task_index} has been completed");
        handler_index -= 1;
        if handler_index == 0 {
            info!("collecting_task: all tasks have been stopped, exiting");
            return Ok(());
        }
    }

    Ok(())
}

async fn collect_one_db_instance(
    database: ScrapeConfigDatabase,
    shutdown_channel: ShutdownReceiver,
) -> Result<(), PsqlExporterError> {
    debug!("collect_one_db_instance: start task for {database:?}");
    let certificates =
        PostgresSslCertificates::from(database.sslrootcert, database.sslcert, database.sslkey)?;
    let mut db_connection = PostgresConnection::new(
        database.connection_string,
        database.sslmode.unwrap(),
        certificates,
        database.backoff_interval,
        database.max_backoff_interval,
        shutdown_channel.clone(),
    )
    .await?;

    let registry = prometheus::default_registry();
    let mut query_metrics: Vec<QueryMetrics> = Vec::with_capacity(database.queries.len());
    let mut sleeper = SleepHelper::from(shutdown_channel.clone());

    for q in database.queries.iter() {
        let metric = QueryMetrics::from(q)?;
        query_metrics.push(metric);
    }

    loop {
        for (query_item, index) in database.queries.iter().zip(0..query_metrics.len()) {
            if query_metrics[index].next_query_time > SystemTime::now() {
                continue;
            }

            let result = db_connection
                .query(&query_item.query, query_item.query_timeout)
                .await;

            match result {
                Ok(result) => {
                    query_metrics[index].register(registry);
                    match &query_item.values {
                        ScrapeConfigValues::ValueFrom(value) => {
                            if let Some(field) = &value.field {
                                update_metrics(
                                    &result,
                                    Some(field),
                                    &query_item.var_labels,
                                    &query_metrics[index].metrics[0],
                                )
                            } else {
                                update_metrics(
                                    &result,
                                    None,
                                    &query_item.var_labels,
                                    &query_metrics[index].metrics[0],
                                )
                            }
                        }
                        ScrapeConfigValues::ValuesWithLabels(values) => {
                            for (value, metric) in values.iter().zip(&query_metrics[index].metrics)
                            {
                                update_metrics(
                                    &result,
                                    Some(&value.field),
                                    &query_item.var_labels,
                                    metric,
                                )
                            }
                        }
                        ScrapeConfigValues::ValuesWithSuffixes(values) => {
                            for (value, metric) in values.iter().zip(&query_metrics[index].metrics)
                            {
                                update_metrics(
                                    &result,
                                    Some(&value.field),
                                    &query_item.var_labels,
                                    metric,
                                )
                            }
                        }
                    }
                }
                Err(e) => {
                    if query_item.metric_expiration_time != Duration::ZERO {
                        let expiration_time =
                            query_metrics[index].last_updated + query_item.metric_expiration_time;
                        if SystemTime::now() > expiration_time {
                            debug!("deregister metrics as expired");
                            query_metrics[index].unregister(registry);
                        }
                    }
                    error!("{e}")
                }
            };
            query_metrics[index].next_query_time = SystemTime::now() + query_item.scrape_interval;
        }

        let next_query_time = query_metrics
            .iter()
            .min_by(|x, y| x.next_query_time.cmp(&y.next_query_time))
            .map(|x| x.next_query_time)
            .expect("looks like a BUG");

        if next_query_time > SystemTime::now() {
            let sleep_time = next_query_time
                .duration_since(SystemTime::now())
                .unwrap_or(Duration::from_micros(0));

            sleeper.sleep(sleep_time).await?;
        }
    }
}

fn update_metrics(
    rows: &[Row],
    field: Option<&str>,
    var_labels: &Option<Vec<String>>,
    metric: &MetricWithType,
) {
    match metric {
        MetricWithType::SingleInt(metric) => {
            if let Some(field) = field {
                metric.set(rows[0].get(field))
            } else {
                metric.set(rows[0].get(0))
            }
        }
        MetricWithType::SingleFloat(metric) => {
            if let Some(field) = field {
                metric.set(rows[0].get(field))
            } else {
                metric.set(rows[0].get(0))
            }
        }
        MetricWithType::VectorInt(metric) => {
            for row in rows {
                let mut new_labels: Vec<String> = vec![];
                if let Some(labels) = var_labels {
                    for label in labels {
                        new_labels.push(row.get(label.as_str()));
                    }
                    let new_labels: Vec<&str> = new_labels.iter().map(AsRef::as_ref).collect();
                    let new_labels: &[&str] = new_labels.as_slice();
                    if let Some(field) = field {
                        metric.with_label_values(new_labels).set(row.get(field));
                    } else {
                        metric.with_label_values(new_labels).set(row.get(0));
                    }
                }
            }
        }
        MetricWithType::VectorFloat(metric) => {
            for row in rows {
                let mut new_labels: Vec<String> = vec![];
                if let Some(labels) = var_labels {
                    for label in labels {
                        new_labels.push(row.get(label.as_str()));
                    }
                    let new_labels: Vec<&str> = new_labels.iter().map(AsRef::as_ref).collect();
                    let new_labels: &[&str] = new_labels.as_slice();
                    if let Some(field) = field {
                        metric.with_label_values(new_labels).set(row.get(field));
                    } else {
                        metric.with_label_values(new_labels).set(row.get(0));
                    }
                }
            }
        }
    }
}
