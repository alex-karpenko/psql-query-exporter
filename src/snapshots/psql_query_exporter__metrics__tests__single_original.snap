---
source: src/metrics.rs
description: "collector test case 'single', original data"
---
# HELP single_with_const_labels single_with_const_labels
# TYPE single_with_const_labels gauge
single_with_const_labels{label_key_1="const_label_value_1"} 3
# HELP single_with_expiration single_with_expiration
# TYPE single_with_expiration gauge
single_with_expiration 0
# HELP single_with_var_labels single_with_var_labels
# TYPE single_with_var_labels gauge
single_with_var_labels{name="Jack"} 3
single_with_var_labels{name="Jane"} 2
single_with_var_labels{name="John"} 1
# HELP single_without_labels single_without_labels
# TYPE single_without_labels gauge
single_without_labels 3
# HELP single_without_labels_with_desc single basic with description
# TYPE single_without_labels_with_desc gauge
single_without_labels_with_desc 3
